### PR TITLE
Feat/Lua loader

### LINKED ISSUE #675: Lua Loader
## 🚀 Feature

Lua is very easy to embed but we may be able to make it simpler and safer. It would be interesting to add it as a supported language, obviously LuaJIT:

https://luajit.org/install.html

### PR DESCRIPTION
**Pull request by k5602**
# Description
This PR implements a robust, production-ready Lua loader for MetaCall, transitioning from standard scaffolded Lua to **LuaJIT**. It introduces a complete value conversion system, enhanced function discovery, and comprehensive edge-case testing.

### Key Changes
* **LuaJIT Integration**: Added `FindLuaJIT.cmake` and updated the build system to prioritize LuaJIT for improved performance and integration.
* **Deep Value Conversion**: Implemented `lua_loader_convert.c` which handles bidirectional conversion between Lua tables (detecting both arrays and maps) and MetaCall types, including a safety recursion limit of 10 as a best practice.
* **Environment Isolation**: Each loaded script now runs in its own isolated environment table, preventing global namespace pollution between handles.
* **Enhanced Discovery**: Function discovery  uses `debug.getinfo` to introspect parameter counts and supports variadic functions (up to 64 arguments).
* **Execution Path Management**: Added support for adding and searching multiple execution paths for relative file lookups.
* **Edge-Case Test Suite**: Introduced `metacall_lua_mixed_test`, a comprehensive test suite covering 40+ scenarios including numeric boundaries (INT64), Unicode strings, sparse arrays, and multiple return values.

Closes #675 

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected) - *Switch to LuaJIT requirement.*
- [ ] This change requires a documentation update
- [x] Documentation update - *Updated READMEs and inline documentation.*

# Checklist:
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests/screenshots (if any) that prove my fix is effective or that my feature works.
- [x] I have tested the tests implicated (if any) by my own code and they pass (`make test` or `ctest -VV -R <test-name>`).
- [ ] If my change is significant or breaking, I have passed all tests with `./docker-compose.sh test &> output` and attached the output.
- [ ] I have tested my code with `OPTION_BUILD_ADDRESS_SANITIZER` or `./docker-compose.sh test-address-sanitizer &> output` and `OPTION_TEST_MEMORYCHECK`.
- [ ] I have tested my code with `OPTION_BUILD_THREAD_SANITIZER` or `./docker-compose.sh test-thread-sanitizer &> output`.
- [ ] I have tested with `Helgrind` in case my code works with threading.
- [x] I have run `make clang-format` in order to format my code and my code follows the style guidelines.

### CODE REVIEW

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 1]
```diff
@@ -331,106 +168,286 @@ loader_impl_data lua_loader_impl_initialize(loader_impl impl, configuration conf
...
+		{
+			loaded = 1;
+		}
+		else
 		{
-			log_write("metacall", LOG_LEVEL_Error, "Lua module %s failed to load: %s", name, lua_tostring(lua_impl->vm, -1));
+			const char *error_text = lua_tostring(lua_impl->vm, -1);
+			if (error_text != NULL)
+			{
+				snprintf(error_message, sizeof(error_message), "%s", error_text);
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c]: why don't you just use error_text instead?

Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c]: it's on the vm stack and later on it's freed? if that's the case strlen it and malloc the string, copy it, then free it.

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 2]
```diff
@@ -331,106 +168,286 @@ loader_impl_data lua_loader_impl_initialize(loader_impl impl, configuration conf
...
+						path = join_path;
+						loaded = 1;
+						break;
+					}
+					else
+					{
+						const char *error_text = lua_tostring(lua_impl->vm, -1);
+						if (error_text != NULL)
+						{
+							snprintf(error_message, sizeof(error_message), "%s", error_text);
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c]: I see, you're always storing the last error, maybe you can just log them in place and avoid copying them...

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 3]
```diff
@@ -442,87 +459,296 @@ int lua_loader_impl_clear(loader_impl impl, loader_handle handle)
...
 }
 
 int lua_loader_impl_destroy(loader_impl impl)
 {
-	loader_impl_lua lua_impl = loader_impl_get(impl);
+	loader_impl_lua lua_impl = (loader_impl_lua)loader_impl_get(impl);
 
 	if (lua_impl != NULL)
 	{
-		/* Destroy children loaders */
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c:518]: leave the comment..

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 4]
```diff
@@ -442,87 +459,296 @@ int lua_loader_impl_clear(loader_impl impl, loader_handle handle)
...
 		loader_unload_children(impl);
 
-		/* Destroy Lua VM */
-		lua_close(lua_impl->vm);
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c:522]: isn't this needed?

File: source/scripts/lua/mixed-test/source/mixed_test.lua [hunk 5]
```diff
@@ -0,0 +1,239 @@
+-- mixed_test.lua
+-- Comprehensive edge case test script for Lua loader
+-- Tests numeric boundaries, string handling, containers, and error conditions
+
+-- Numeric boundary functions
```
Comment (viferga) [inline source/scripts/lua/mixed-test/source/mixed_test.lua:5]: this must go inside scripts folder

File: source/tests/metacall_lua_mixed_test/README.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
+# Lua Loader Mixed Edge Case Tests
+
+## Test Coverage
+
+### 1. Numeric Boundary Tests (6 tests)
```
Comment (viferga) [inline source/tests/metacall_lua_mixed_test/README.md:10]: this can be part of the loader doc itself, no? because we are testing details of the language itself maybe?

File: source/scripts/lua/test/source/test.lua [hunk 7]
```diff
@@ -0,0 +1,74 @@
+-- Test script for Lua loader
+-- Contains simple functions for testing type handling, arrays, and error conditions
+
+-- Function returning multiple types for testing
```
Comment (viferga) [inline source/scripts/lua/test/source/test.lua:4]: this must go in scripts too

File: source/tests/metacall_lua_test/CMakeLists.txt [hunk 8]
```diff
@@ -52,6 +52,12 @@ add_executable(${target}
 	${sources}
 )
 
+configure_file(
```
Comment (viferga) [inline source/tests/metacall_lua_test/CMakeLists.txt]: this must be done with scripts infrastructure, not like this

File: source/tests/metacall_lua_test/CMakeLists.txt [hunk 9]
```diff
@@ -95,6 +101,7 @@ target_link_libraries(${target}
 target_compile_definitions(${target}
 	PRIVATE
 	${DEFAULT_COMPILE_DEFINITIONS}
+	LUA_LOADER_TEST_SCRIPT_ABSOLUTE_PATH="${LOADER_SCRIPT_PATH}/test.lua"
```
Comment (viferga) [inline source/tests/metacall_lua_test/CMakeLists.txt]: this should be removed, with the scripts infrastructure the scripts get copied into the scripts path and that path os passed to execution_path at startup.. so it's not needed to be passed separately

**Review by viferga — COMMENTED** [review]
great job man

File: source/loaders/lua_loader/source/lua_loader_impl.c (hunk 2 above)
Comment (viferga) [reply source/loaders/lua_loader/source/lua_loader_impl.c]: well this may not work if it tries few paths that fail and one works at the end, it will be confusing for the end user...

File: source/loaders/lua_loader/source/lua_loader_impl.c (hunk 4 above)
Comment (k5602) [reply source/loaders/lua_loader/source/lua_loader_impl.c:522]: yes needed and present , i will revert this

File: source/loaders/lua_loader/source/lua_loader_impl.c (hunk 1 above)
Comment (k5602) [reply source/loaders/lua_loader/source/lua_loader_impl.c]: yes i will update it for truncation sake.

File: source/tests/metacall_lua_mixed_test/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+	COMMAND $<TARGET_FILE:${target}>
+)
+
+#
+# Define dependencies
+#
+
+add_dependencies(${target}
+	lua_loader
+	lua-mixed_test
```
Comment (viferga) [inline source/tests/metacall_lua_mixed_test/CMakeLists.txt]: what is this?  can you use lua-mixed-test?

File: source/tests/metacall_lua_test/source/metacall_lua_test.cpp [hunk 11]
```diff
@@ -26,7 +26,7 @@
 
 class metacall_lua_test : public testing::Test
 {
-public:
```
Comment (viferga) [inline source/tests/metacall_lua_test/source/metacall_lua_test.cpp:29]: this was fine

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 12]
```diff
@@ -331,106 +168,287 @@ loader_impl_data lua_loader_impl_initialize(loader_impl impl, configuration conf
...
+static void lua_loader_impl_error_capture(lua_State *L, char *error_message, size_t error_message_size)
+{
+	size_t len = 0;
+	const char *error_text = lua_tolstring(L, -1, &len);
 
-		return (loader_handle)handle;
-	}
+	/* Copy the error before popping it from the Lua VM stack */
+	snprintf(error_message, error_message_size, "%.*s", (int)len,
+		error_text != NULL ? error_text : "unknown error");
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c]: just use strncpy here

File: source/loaders/lua_loader/source/lua_loader_impl.c [hunk 13]
```diff
@@ -331,106 +168,287 @@ loader_impl_data lua_loader_impl_initialize(loader_impl impl, configuration conf
...
 
-loader_handle lua_loader_impl_load_from_memory(loader_impl impl, const loader_name name, const char *buffer, size_t size, void *data)
+loader_handle lua_loader_impl_load_from_file(loader_impl impl, const loader_path paths[], size_t size, void *data)
 {
-	loader_impl_lua lua_impl = loader_impl_get(impl);
-	loader_impl_lua_handle handle = malloc(sizeof(struct loader_impl_lua_handle_type));
+	loader_impl_lua lua_impl = (loader_impl_lua)loader_impl_get(impl);
+	loader_impl_lua_handle handle;
+	size_t iterator;
+	char error_message[512];
```
Comment (viferga) [inline source/loaders/lua_loader/source/lua_loader_impl.c]: allocate this in the heap if you can, with dynamic size, use an inout pointer if you need so in the capture functiob

Comment (viferga) [reply source/loaders/lua_loader/source/lua_loader_impl.c]: the error_message I mean

**Comment by viferga** [comment]
There is still many things wrong:

1) Find script is not coherent with versions.
2) Install script is not available.
3) Tools for environemnt.sh/ps1.
4) Solve the error handling part of the JIT and running the tests.

I am going to merge it because it is new code but otherwise I won't do it. I will work on it today for solving the issues.

**Comment by viferga** [comment]
Full review:
https://github.com/metacall/core/commit/e54e3c81e2859a8985ba6d36a3b47b1b5cf9ea05
https://github.com/metacall/core/commit/f87d8e9323ca29c1db389ed043e2376aa9c1ac52
https://github.com/metacall/core/commit/42710112a6a54d31ca764908564eeef50562cbea
https://github.com/metacall/core/commit/484b1c7e63222e3757b6932864e78e4aa11260ba
https://github.com/metacall/core/commit/9dc171966fd4872be9862f0a9e4319ca6e8a1ca5
https://github.com/metacall/core/commit/625d716889751c1a48b7dcb952dc9df0ff49f7b0
https://github.com/metacall/core/commit/dac3364ddc6e62cb7b4f5d5049f6c6901ebb69a7
